File: client/myna-core/src/audio.rs
//! Audio types, mirroring Python `myna.core.audio` and the Rust sketch in
//! `docs/audio-adapter-api.md` §2.
//!
//! Under the audio-push invariant the *client* owns capture and pushes PCM; the
//! STT service never touches the microphone. PCM is carried as raw binary WS
//! frames, not JSON — so [`PcmChunk`] is deliberately not `Serialize`. Only
//! [`AudioFormat`] crosses the wire as JSON (inside the session config), so only
//! it derives serde.

use std::time::Duration;

use bytes::Bytes;
use serde::{Deserialize, Serialize};

/// Raw PCM stream description. Default is 16 kHz mono S16LE — the common
/// denominator across the candidate ASR models.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioFormat {
    pub sample_rate_hz: u32,
    pub channels: u8,
    /// Signed little-endian integer sample width. 2 = S16LE.
    pub sample_width_bytes: u8,
}

impl Default for AudioFormat {
    fn default() -> Self {
        Self {
            sample_rate_hz: 16_000,
            channels: 1,
            sample_width_bytes: 2,
        }
    }
}

impl AudioFormat {
    pub fn bytes_per_second(&self) -> u32 {
        self.sample_rate_hz * self.channels as u32 * self.sample_width_bytes as u32
    }
}

/// A contiguous slice of raw PCM in `format`. `Bytes` is cheap to clone so a
/// chunk can fan out to a level meter without copying (audio-adapter API §2/§8).
#[derive(Clone, Debug)]
pub struct PcmChunk {
    pub data: Bytes,
    pub format: AudioFormat,
}

impl PcmChunk {
    pub fn new(data: impl Into<Bytes>, format: AudioFormat) -> Self {
        Self {
            data: data.into(),
            format,
        }
    }

    pub fn duration(&self) -> Duration {
        let bps = self.format.bytes_per_second();
        if bps == 0 {
            return Duration::ZERO;
        }
        Duration::from_secs_f64(self.data.len() as f64 / bps as f64)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_format_matches_python() {
        let fmt = AudioFormat::default();
        assert_eq!(fmt.sample_rate_hz, 16_000);
        assert_eq!(fmt.channels, 1);
        assert_eq!(fmt.sample_width_bytes, 2);
        assert_eq!(fmt.bytes_per_second(), 32_000);
    }

    #[test]
    fn audio_format_wire_shape() {
        let expected: serde_json::Value = serde_json::from_str(
            r#"{"sample_rate_hz": 16000, "channels": 1, "sample_width_bytes": 2}"#,
        )
        .unwrap();
        assert_eq!(
            serde_json::to_value(AudioFormat::default()).unwrap(),
            expected
        );
    }

    #[test]
    fn duration_from_one_second_of_audio() {
        let fmt = AudioFormat::default();
        let chunk = PcmChunk::new(vec![0u8; fmt.bytes_per_second() as usize], fmt);
        assert_eq!(chunk.duration(), Duration::from_secs(1));
    }
}

File: server/src/myna/core/audio.py
"""Audio types shared by the desktop audio pipeline and the testbed.

The audio-push model (see CLAUDE.md) means the *client* owns capture: PCM is
produced client-side (PipeWire on the desktop, scripted/file sources in the
testbed) and pushed to the STT service. The service never touches PipeWire.
"""

from __future__ import annotations

import logging
from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import Protocol, runtime_checkable

_log = logging.getLogger(__name__)


@dataclass(frozen=True)
class AudioFormat:
    """Raw PCM stream description. Default is 16 kHz mono S16LE, the common
    denominator for the candidate models."""

    sample_rate_hz: int = 16_000
    channels: int = 1
    sample_width_bytes: int = 2  # signed little-endian integer samples

    @property
    def frame_bytes(self) -> int:
        """Bytes of one sample frame: a sample for every channel."""
        return self.channels * self.sample_width_bytes

    @property
    def bytes_per_second(self) -> int:
        return self.sample_rate_hz * self.frame_bytes


@dataclass(frozen=True)
class PcmChunk:
    """A contiguous slice of raw PCM audio."""

    data: bytes
    format: AudioFormat

    @property
    def duration_seconds(self) -> float:
        return len(self.data) / self.format.bytes_per_second


class PcmFramer:
    """Cuts a byte stream into whole sample frames. A frame split across
    appends is carried to the next one; at the utterance boundary a partial
    frame cannot be completed, so ``flush`` discards it, loudly, rather than
    pad it or let it straddle into the next utterance."""

    def __init__(self, frame_bytes: int) -> None:
        if frame_bytes < 1:
            raise ValueError(f"a sample frame is at least one byte, got {frame_bytes}")
        self._frame_bytes = frame_bytes
        self._pending = b""

    def feed(self, data: bytes) -> bytes:
        if self._pending:
            data = self._pending + data
        whole = len(data) - len(data) % self._frame_bytes
        self._pending = data[whole:]
        return data[:whole]

    def flush(self) -> None:
        if self._pending:
            _log.warning(
                "discarding %d byte(s) of a partial PCM sample frame at the utterance boundary",
                len(self._pending),
            )
            self._pending = b""


@runtime_checkable
class AudioSource(Protocol):
    """Produces PCM chunks. Implementations control pacing: a real-time source
    sleeps between chunks to mimic live capture; a batch source does not.

    Audio is never persisted — sources stream from memory, a capture stack,
    or read-only test fixtures.
    """

    @property
    def format(self) -> AudioFormat: ...

    def chunks(self) -> AsyncIterator[PcmChunk]: ...
